### PR TITLE
SAK-47674: ROSTER: Improving method getAndCacheSortedMembership

### PR DESCRIPTION
**Pull request by victorGomollon**
[SAK-47674: ROSTER: Improving method getAndCacheSortedMembership](https://sakaiproject.atlassian.net/browse/SAK-47674)

### CODE REVIEW

File: roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java [hunk 1]
```diff
@@ -915,65 +912,41 @@ private List<RosterMember> getAndCacheSortedMembership(Site site, String groupId
...
-
-                // Same for site#role
-                roles.forEach(r -> cache.put(siteId + "#" + r.getId(), new ArrayList<RosterMember>()));
-
-                for (Member member : membership) {
-                    try {
-                        RosterMember rosterMember = getRosterMember(userMap, groups, member, site, pronunceMap);
-
-                        siteMembers.add(rosterMember);
+            Map<String, List<RosterMember>> cacheMembersMap = new HashMap<String, List<RosterMember>>();
```
Comment (adrianfish) [inline roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java:915]: You don't need to add the types twice here. Just on the Map will work.

File: roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java [hunk 2]
```diff
@@ -915,65 +912,41 @@ private List<RosterMember> getAndCacheSortedMembership(Site site, String groupId
...
-
-                        List<RosterMember> roleMembers = (List<RosterMember>) cache.get(siteId + "#" + memberRoleId);
-                        roleMembers.add(rosterMember);
-                    } catch (UserNotDefinedException e) {
-                        log.warn("user not found: " + e.getId());
-                    }
-                }
+            for (Role role : roles) {
+                cacheMembersMap.put(siteId + "#" + role.getId(), new ArrayList<RosterMember>());
+                groups.forEach(group -> cacheMembersMap.put(siteId + "#" + group.getId() + "#" + role.getId(), new ArrayList<RosterMember>()));
```
Comment (adrianfish) [inline roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java:919]: You don't need the type on the ArrayList

File: roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java [hunk 3]
```diff
@@ -915,65 +912,41 @@ private List<RosterMember> getAndCacheSortedMembership(Site site, String groupId
...
-                // Cache the main site list
-                cache.put(siteId, siteMembers);
+			cacheMembersMap.put(siteId, siteMembers);
+			log.debug("Caching on '{}' ...", siteId);
 
-                return new ArrayList<>((List<RosterMember>) cache.get(key));
-            }
+			cacheMembersMap.values().forEach(a -> Collections.sort(a, memberComparator));
+			cache.putAll(cacheMembersMap);
+            return (List<RosterMember>) cache.get(key);
```
Comment (adrianfish) [inline roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java:946]: Indentation is off here.

**Review by adrianfish — COMMENTED** [review]
Just some formatting. I like the changes. I assume that you can remove the synchronized because you're just setting the cache entry in one line of code?

File: roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java (hunk 2 above)
Comment (victorGomollon) [reply roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java]: done

File: roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java (hunk 3 above)
Comment (victorGomollon) [reply roster2/tool/src/java/org/sakaiproject/roster/impl/SakaiProxyImpl.java]: done

**Comment by adrianfish** [comment]
I do have an overall question about this PR. The synchronized block is there to prevent multiple concurrent hits on some expensive parts of Sakai, like the UserDirectoryService. So, even though you're putting into the cache in one line, you're now allowing a race condition on some pretty heavy api calls. So, have you turned on hibernate query logging and run this through a profiler like YourKit? I would if it was me, because this method is *the* bottleneck of the roster tool.

**Comment by ottenhoff** [comment]
This was a long time ago, but we saw lots of ConcurrentModification exceptions back in the 11.x days:

https://sakaiproject.atlassian.net/browse/SAK-33322

**Comment by SedueRey** [comment]
Hi!!
@adrianfish we are going to check it in our instances. Give us some time, we'll tell you something as fast as we can.
Thanks!

**Review by juanjmerono — APPROVED** [review]

**Comment by adrianfish** [comment]
Great, good to hear. Thanks for checking that out.

**Pull request merged**